File: core/src/commonTest/kotlin/sudoku/core/SolverTest.kt
package sudoku.core

import sudoku.core.model.*
import sudoku.core.solver.SimpleSolver
import sudoku.core.solver.Solver
import sudoku.core.solver.StepFinder
import kotlin.test.*

class SolverTest {
    @Test
    fun testSolveEasyPuzzle() {
        // This puzzle is solvable with singles only
        val puzzle = "530070000600195000098000060800060003400803001700020006060000280000419005000080079"
        val board = Board()
        board.loadFromString(puzzle)
        board.solutionSet = false

        val solver = Solver()
        val result = solver.solve(board)
        assertTrue(result.solved)
        // Should use only singles techniques
        for (step in result.steps) {
            assertTrue(
                step.type == SolutionType.FULL_HOUSE ||
                    step.type == SolutionType.NAKED_SINGLE ||
                    step.type == SolutionType.HIDDEN_SINGLE,
                "Easy puzzle should only use singles, got: ${step.type}",
            )
        }
    }

    @Test
    fun testSimpleSolverFindsSingles() {
        val board = Board()
        // Set up row 0 with 8 of 9 digits
        for (i in 0..7) {
            board.setCell(i, i + 1, isFixed = true)
        }
        board.setAllExposedSingles()

        // If cell 8 wasn't auto-set by singles propagation, the solver should find it
        if (board.values[8] == 0) {
            val solver = SimpleSolver()
            val steps = solver.findSteps(board)
            assertTrue(steps.isNotEmpty(), "Should find a single")
            assertEquals(9, steps[0].value)
        } else {
            // It was already set by propagation, which is fine
            assertEquals(9, board.values[8])
        }
    }

    @Test
    fun testDifficultyFromScore() {
        assertEquals(Difficulty.EASY, Difficulty.fromScore(100))
        assertEquals(Difficulty.EASY, Difficulty.fromScore(799))
        assertEquals(Difficulty.MEDIUM, Difficulty.fromScore(800))
        assertEquals(Difficulty.MEDIUM, Difficulty.fromScore(999))
        assertEquals(Difficulty.HARD, Difficulty.fromScore(1000))
        assertEquals(Difficulty.HARD, Difficulty.fromScore(1599))
        assertEquals(Difficulty.UNFAIR, Difficulty.fromScore(1600))
        assertEquals(Difficulty.EXTREME, Difficulty.fromScore(1800))
        assertEquals(Difficulty.EXTREME, Difficulty.fromScore(10000))
    }

    @Test
    fun testDifficultyEnumComparison() {
        assertTrue(Difficulty.EASY < Difficulty.MEDIUM)
        assertTrue(Difficulty.MEDIUM < Difficulty.HARD)
        assertTrue(Difficulty.HARD < Difficulty.UNFAIR)
        assertTrue(Difficulty.UNFAIR < Difficulty.EXTREME)
        assertTrue(Difficulty.EASY <= Difficulty.EASY)
    }

    @Test
    fun testSolutionTypeDifficultyConsistency() {
        // Singles should be EASY
        assertEquals(Difficulty.EASY, SolutionType.FULL_HOUSE.difficulty)
        assertEquals(Difficulty.EASY, SolutionType.NAKED_SINGLE.difficulty)
        assertEquals(Difficulty.EASY, SolutionType.HIDDEN_SINGLE.difficulty)

        // Locked candidates and basic subsets should be MEDIUM
        assertEquals(Difficulty.MEDIUM, SolutionType.LOCKED_CANDIDATES_1.difficulty)
        assertEquals(Difficulty.MEDIUM, SolutionType.NAKED_PAIR.difficulty)
        assertEquals(Difficulty.MEDIUM, SolutionType.HIDDEN_TRIPLE.difficulty)

        // Fish and wings should be HARD
        assertEquals(Difficulty.HARD, SolutionType.X_WING.difficulty)
        assertEquals(Difficulty.HARD, SolutionType.XY_WING.difficulty)
        assertEquals(Difficulty.HARD, SolutionType.NAKED_QUADRUPLE.difficulty)

        // Coloring should be UNFAIR
        assertEquals(Difficulty.UNFAIR, SolutionType.SIMPLE_COLORS_TRAP.difficulty)
        assertEquals(Difficulty.UNFAIR, SolutionType.MULTI_COLORS_1.difficulty)

        // Chains and brute force should be EXTREME
        assertEquals(Difficulty.EXTREME, SolutionType.X_CHAIN.difficulty)
        assertEquals(Difficulty.EXTREME, SolutionType.BRUTE_FORCE.difficulty)
    }

    @Test
    fun testSolutionTypeDifficultyMonotonicity() {
        // Within each category, technique difficulty should not decrease
        // (i.e., no HARD technique in a category that also has EASY techniques
        //  should appear before the EASY ones in enum order)
        // More practically: every technique's difficulty should be >= EASY
        for (type in SolutionType.entries) {
            assertTrue(
                type.difficulty >= Difficulty.EASY,
                "${type.name} difficulty should be at least EASY",
            )
        }
    }

    @Test
    fun testGatedSolverEasyPuzzle() {
        // The known easy puzzle should solve with maxDifficulty=EASY
        val puzzle = "530070000600195000098000060800060003400803001700020006060000280000419005000080079"
        val board = Board()
        board.loadFromString(puzzle)
        board.solutionSet = false

        val solver = Solver()
        val result = solver.solve(board, maxDifficulty = Difficulty.EASY)
        assertTrue(result.solved, "Easy puzzle should be solvable with EASY gate")
    }

    @Test
    fun testGatedSolverRejectsAboveGate() {
        // Solve the easy puzzle ungated first - it should succeed
        val puzzle = "530070000600195000098000060800060003400803001700020006060000280000419005000080079"
        val board = Board()
        board.loadFromString(puzzle)

        val solver = Solver()

        // Now test: ungated solve to find what techniques are used
        val ungatedResult = solver.solve(board)
        assertTrue(ungatedResult.solved)

        // If the puzzle uses any non-EASY technique in ungated mode,
        // gating to EASY should fail. But this puzzle is pure singles,
        // so let's verify the gating mechanism works by checking that
        // a null maxDifficulty behaves the same as no gating.
        val nullGatedResult = solver.solve(board, maxDifficulty = null)
        assertEquals(ungatedResult.solved, nullGatedResult.solved)
        assertEquals(ungatedResult.score, nullGatedResult.score)
    }

    @Test
    fun testGatedSolverStopsWhenStuck() {
        // A harder puzzle that requires more than singles
        // This is a well-known puzzle requiring locked candidates / pairs
        val puzzle = "000000010400000000020000000000050407008000300001090000300400200050100000000806000"
        val board = Board()
        board.loadFromString(puzzle)

        val solver = Solver()

        // Ungated should solve
        val ungatedResult = solver.solve(board)
        assertTrue(ungatedResult.solved, "Should be solvable ungated")

        // Check if it uses any technique above EASY
        val usesAdvanced = ungatedResult.steps.any { it.type.difficulty > Difficulty.EASY }
        if (usesAdvanced) {
            // Gating to EASY should fail (get stuck)
            val gatedResult = solver.solve(board, maxDifficulty = Difficulty.EASY)
            assertFalse(gatedResult.solved, "Should not solve with EASY gate when advanced techniques are needed")
        }
    }

    // ── Locked Subset Tests ──────────────────────────────────────────────────

    private fun findStepOfType(
        puzzleString: String,
        targetType: SolutionType,
    ): Pair<Board, SolutionStep>? {
        val stepFinder = StepFinder()
        val board = Board()
        board.loadFromString(puzzleString)

        while (!board.isSolved) {
            val step = stepFinder.findNextStep(board) ?: break
            if (step.type == targetType) return board to step

            if (step.type.isSingle || step.type == SolutionType.BRUTE_FORCE) {
                board.setCell(step.cellIndex, step.value)
                board.setAllExposedSingles()
            } else {
                for ((cellIndex, candidate) in step.candidatesRemoved) {
                    board.setCandidate(cellIndex, candidate, false)
                }
                board.setAllExposedSingles()
            }
        }
        return null
    }

    @Test
    fun testSolverFindsLockedPair() {
        val puzzles =
            listOf(
                "000000010400000000020000000000050407008000300001090000300400200050100000000806000",
                "100007090030020008009600500005300900010080002600004000300000010040000007007000300",
                "003000200090000006000630040006003900050904020004100800020048000400000050005000300",
            )
        var found: Pair<Board, SolutionStep>? = null
        for (puzzle in puzzles) {
            found = findStepOfType(puzzle, SolutionType.LOCKED_PAIR)
            if (found != null) break
        }
        if (found == null) return // Skip if no locked pair found in curated puzzles

        val (board, step) = found
        assertEquals(SolutionType.LOCKED_PAIR, step.type)
        assertEquals(2, step.indices.size, "Locked pair should have 2 cells")
        assertTrue(step.candidatesRemoved.isNotEmpty(), "Should have eliminations")

        // Verify all subset cells share the same block and line
        val blocks = step.indices.map { Board.getBlock(it) }.toSet()
        assertEquals(1, blocks.size, "All cells must be in the same block")
        val rows = step.indices.map { it / 9 }.toSet()
        val cols = step.indices.map { it % 9 }.toSet()
        assertTrue(rows.size == 1 || cols.size == 1, "All cells must be in the same row or col")

        // Combined candidates should be exactly 2
        var combinedMask = 0
        for (cell in step.indices) combinedMask = combinedMask or board.cells[cell]
        assertEquals(2, Board.ANZ_VALUES[combinedMask], "Combined candidates should be exactly 2")
    }

    @Test
    fun testSolverFindsLockedTriple() {
        val puzzles =
            listOf(
                "000000010400000000020000000000050407008000300001090000300400200050100000000806000",
                "100007090030020008009600500005300900010080002600004000300000010040000007007000300",
                "003000200090000006000630040006003900050904020004100800020048000400000050005000300",
            )
        var found: Pair<Board, SolutionStep>? = null
        for (puzzle in puzzles) {
            found = findStepOfType(puzzle, SolutionType.LOCKED_TRIPLE)
            if (found != null) break
        }
        if (found == null) return // Skip if not found in curated puzzles

        val (board, step) = found
        assertEquals(SolutionType.LOCKED_TRIPLE, step.type)
        assertEquals(3, step.indices.size, "Locked triple should have 3 cells")
        assertTrue(step.candidatesRemoved.isNotEmpty(), "Should have eliminations")

        val blocks = step.indices.map { Board.getBlock(it) }.toSet()
        assertEquals(1, blocks.size, "All cells must be in the same block")

        var combinedMask = 0
        for (cell in step.indices) combinedMask = combinedMask or board.cells[cell]
        assertEquals(3, Board.ANZ_VALUES[combinedMask], "Combined candidates should be exactly 3")
    }

    @Test
    fun testLockedPairHasSolver() {
        assertTrue(SolutionType.LOCKED_PAIR.hasSolver, "LOCKED_PAIR should have solver")
        assertTrue(SolutionType.LOCKED_TRIPLE.hasSolver, "LOCKED_TRIPLE should have solver")
    }
}

File: rust/sudoku-core/tests/solver_test.rs
use sudoku_core::board::Board;
use sudoku_core::solver::{SolverOrchestrator, StepFinder};
use sudoku_core::types::{Difficulty, SolutionType};
use sudoku_core::step::SolutionStep;

const EASY_PUZZLE: &str = "530070000600195000098000060800060003400803001700020006060000280000419005000080079";

#[test]
fn test_solve_easy_puzzle() {
    let mut board = Board::new();
    board.load_from_string(EASY_PUZZLE);
    board.solution_set = false;

    let solver = SolverOrchestrator::new();
    let result = solver.solve(&board, None);
    assert!(result.solved);
    for step in &result.steps {
        assert!(
            step.step_type == SolutionType::FullHouse
                || step.step_type == SolutionType::NakedSingle
                || step.step_type == SolutionType::HiddenSingle,
            "Easy puzzle should only use singles, got: {:?}",
            step.step_type
        );
    }
}

#[test]
fn test_difficulty_from_score() {
    assert_eq!(Difficulty::from_score(100), Difficulty::Easy);
    assert_eq!(Difficulty::from_score(799), Difficulty::Easy);
    assert_eq!(Difficulty::from_score(800), Difficulty::Medium);
    assert_eq!(Difficulty::from_score(999), Difficulty::Medium);
    assert_eq!(Difficulty::from_score(1000), Difficulty::Hard);
    assert_eq!(Difficulty::from_score(1599), Difficulty::Hard);
    assert_eq!(Difficulty::from_score(1600), Difficulty::Unfair);
    assert_eq!(Difficulty::from_score(1800), Difficulty::Extreme);
    assert_eq!(Difficulty::from_score(10000), Difficulty::Extreme);
}

#[test]
fn test_difficulty_enum_comparison() {
    assert!(Difficulty::Easy < Difficulty::Medium);
    assert!(Difficulty::Medium < Difficulty::Hard);
    assert!(Difficulty::Hard < Difficulty::Unfair);
    assert!(Difficulty::Unfair < Difficulty::Extreme);
    assert!(Difficulty::Easy <= Difficulty::Easy);
}

#[test]
fn test_solution_type_difficulty_consistency() {
    assert_eq!(SolutionType::FullHouse.difficulty(), Difficulty::Easy);
    assert_eq!(SolutionType::NakedSingle.difficulty(), Difficulty::Easy);
    assert_eq!(SolutionType::HiddenSingle.difficulty(), Difficulty::Easy);

    assert_eq!(SolutionType::LockedCandidates1.difficulty(), Difficulty::Medium);
    assert_eq!(SolutionType::NakedPair.difficulty(), Difficulty::Medium);
    assert_eq!(SolutionType::HiddenTriple.difficulty(), Difficulty::Medium);

    assert_eq!(SolutionType::XWing.difficulty(), Difficulty::Hard);
    assert_eq!(SolutionType::XyWing.difficulty(), Difficulty::Hard);
    assert_eq!(SolutionType::NakedQuadruple.difficulty(), Difficulty::Hard);

    assert_eq!(SolutionType::SimpleColorsTrap.difficulty(), Difficulty::Unfair);
    assert_eq!(SolutionType::MultiColors1.difficulty(), Difficulty::Unfair);

    assert_eq!(SolutionType::XChain.difficulty(), Difficulty::Extreme);
    assert_eq!(SolutionType::BruteForce.difficulty(), Difficulty::Extreme);
}

#[test]
fn test_gated_solver_easy_puzzle() {
    let mut board = Board::new();
    board.load_from_string(EASY_PUZZLE);
    board.solution_set = false;

    let solver = SolverOrchestrator::new();
    let result = solver.solve(&board, Some(Difficulty::Easy));
    assert!(result.solved, "Easy puzzle should be solvable with EASY gate");
}

#[test]
fn test_gated_solver_stops_when_stuck() {
    let puzzle = "000000010400000000020000000000050407008000300001090000300400200050100000000806000";
    let mut board = Board::new();
    board.load_from_string(puzzle);

    let solver = SolverOrchestrator::new();

    let ungated = solver.solve(&board, None);
    assert!(ungated.solved, "Should be solvable ungated");

    let uses_advanced = ungated.steps.iter().any(|s| s.step_type.difficulty() > Difficulty::Easy);
    if uses_advanced {
        let gated = solver.solve(&board, Some(Difficulty::Easy));
        assert!(!gated.solved, "Should not solve with EASY gate when advanced techniques needed");
    }
}

// Helper: replay solver to find a specific technique type
fn find_step_of_type(puzzle: &str, target: SolutionType) -> Option<(Board, SolutionStep)> {
    let finder = StepFinder::new();
    let mut board = Board::new();
    board.load_from_string(puzzle);

    while !board.is_solved() {
        let step = finder.find_next_step(&board, None)?;
        if step.step_type == target {
            return Some((board, step));
        }

        if step.step_type.is_single() || step.step_type == SolutionType::BruteForce {
            board.set_cell(step.cell_index as usize, step.value, false);
            board.set_all_exposed_singles();
        } else {
            for &(cell_index, candidate) in &step.candidates_removed {
                board.set_candidate(cell_index, candidate, false);
            }
            board.set_all_exposed_singles();
        }
    }
    None
}

#[test]
fn test_solver_finds_locked_pair() {
    let puzzles = [
        "000000010400000000020000000000050407008000300001090000300400200050100000000806000",
        "100007090030020008009600500005300900010080002600004000300000010040000007007000300",
        "003000200090000006000630040006003900050904020004100800020048000400000050005000300",
    ];
    let mut found = None;
    for puzzle in &puzzles {
        found = find_step_of_type(puzzle, SolutionType::LockedPair);
        if found.is_some() { break; }
    }
    if let Some((board, step)) = found {
        assert_eq!(step.step_type, SolutionType::LockedPair);
        assert_eq!(step.indices.len(), 2, "Locked pair should have 2 cells");
        assert!(!step.candidates_removed.is_empty(), "Should have eliminations");

        // All in same block
        let blocks: std::collections::HashSet<_> = step.indices.iter()
            .map(|&i| sudoku_core::tables::get_block(i))
            .collect();
        assert_eq!(blocks.len(), 1, "All cells must be in the same block");

        // All in same row or col
        let rows: std::collections::HashSet<_> = step.indices.iter().map(|&i| i / 9).collect();
        let cols: std::collections::HashSet<_> = step.indices.iter().map(|&i| i % 9).collect();
        assert!(rows.len() == 1 || cols.len() == 1, "All cells must be in same row or col");

        // Combined candidates = exactly 2
        let mut combined = 0u16;
        for &cell in &step.indices {
            combined |= board.candidates[cell];
        }
        assert_eq!(
            sudoku_core::tables::ANZ_VALUES[combined as usize], 2,
            "Combined candidates should be exactly 2"
        );
    }
}

#[test]
fn test_locked_pair_has_solver() {
    assert!(SolutionType::LockedPair.has_solver());
    assert!(SolutionType::LockedTriple.has_solver());
}

#[test]
fn test_has_solver_flags() {
    assert!(SolutionType::Skyscraper.has_solver());
    assert!(SolutionType::TwoStringKite.has_solver());
    assert!(SolutionType::EmptyRectangle.has_solver());
    assert!(SolutionType::TurbotFish.has_solver());
}
